### PR TITLE
Improved git resource performance

### PR DESCRIPTION
**Pull request by JulienTD**
## Description

This pull request improves the git resource's performance by getting the hash of only modified files between tags.

## Related Issue(s)
 
N/A

### CODE REVIEW

File: lib/resources/git_resource.py [hunk 1]
```diff
@@ -27,90 +30,159 @@ def __init__(self, database: DbConnector):
...
+    def clone_checkout_and_compute_hashs(self, session_scope, url: str):
         """
-        This method clones the repository from url, retrieves and stores tags in the database,
-        then checkout each tag, sotres files information in the database and computes the hashs.
+        This method clones the repository from url, retrieves tags, compares each tags to retrieve only modified files,
+        computes their hashes and then stores the tags & files information in the database.
         """
-        technology = url.split('.git')[0].split('/')[-1]
         with tempfile.TemporaryDirectory() as tmp_dir_name:
+            technology = url.split('.git')[0].split('/')[-1]
```
Comment (bretfourbe) [inline lib/resources/git_resource.py:158]: `technology` is not needed in this `with` block. Maybe you can move this before the next `with` block  ?

File: lib/resources/git_resource.py [hunk 2]
```diff
@@ -27,90 +30,159 @@ def __init__(self, database: DbConnector):
...
+            commit_diff: DiffIndex = None
+
+            if idx + 1 < len(tag_commit_list):
+                commit_diff = commit.diff(tag_commit_list[idx + 1][0])
+                logger.debug(f"Comparing {tag.name} with {tag_commit_list[idx + 1][1].name}")
+            else:
+                commit_diff = commit.diff(tag_commit_list[idx - 1][0])
+                logger.debug(f"Comparing {tag.name} with {tag_commit_list[idx - 1][1].name}")
+
+            for diffidx in commit_diff:
```
[the remark this answers is not in the document]
Comment (JulienTD) [reply lib/resources/git_resource.py:120]: Done

File: lib/resources/git_resource.py [hunk 3]
```diff
@@ -27,90 +30,159 @@ def __init__(self, database: DbConnector):
...
         This method computes the hash of the provided file and returns it.
         """
         try:
-            with open(file, 'rb') as file_descriptor:
-                file_bytes = file_descriptor.read() # read entire file as bytes
-                readable_hash = hashlib.sha256(file_bytes).hexdigest()
-                return readable_hash
+            readable_hash = hashlib.sha256(file_content).hexdigest()
+            return readable_hash
         except OSError as error:
```
[the remark this answers is not in the document]
Comment (JulienTD) [reply lib/resources/git_resource.py:48]: Indeed, It is not necessary anymore :+1:

File: lib/resources/git_resource.py [hunk 4]
```diff
@@ -27,90 +30,156 @@ def __init__(self, database: DbConnector):
...
+        stack = [commit.tree]
+
+        while len(stack) > 0:
+            tree = stack.pop()
+            # enumerate blobs (files) at this level
+            for blob in tree.blobs:
+                file_list.append((blob.path, blob.hexsha))
+            for subtree in tree.trees:
+                stack.append(subtree)
+        return file_list
```
Comment (polyedre) [inline lib/resources/git_resource.py:64]: Il y a la fonction `git.objects.fun.traverse_tree_recursive` qui fait exactement ça :)

Comment (polyedre) [reply lib/resources/git_resource.py:64]: https://programtalk.com/python-examples/git.objects.fun.traverse_tree_recursive/

Comment (JulienTD) [reply lib/resources/git_resource.py]: J'ai utilisé la fonction `commit.tree.traverse()` à la place parce qu'elle donne le blob hash sous format string et non binaire, de plus, elle est plus facile à utiliser parce qu'elle est déjà présente dans l'objet `commit` :rocket:

File: lib/resources/git_resource.py [hunk 5]
```diff
@@ -27,90 +37,181 @@ def __init__(self, database: DbConnector):
...
+
+        for blob in commit.tree.traverse():
+            if S_ISDIR(blob.mode):
+                continue
+            file_list.append((blob.path, blob.hexsha))
+        return file_list
+
+    def _hash_files(
+        self,
+        files: List[Tuple[FilePath, TagName, BlobHash]],
```
Comment (polyedre) [inline lib/resources/git_resource.py:71]: In order to increase readability, you could use a named dictionary or a dataclass to name objects represented with `Tuple[FilePath, TagName, BlobHash]` and `Tuple[FilePath, TagName, FileHash]` as these two structures are used multiple times.

File: lib/resources/git_resource.py [hunk 6]
```diff
@@ -27,90 +37,181 @@ def __init__(self, database: DbConnector):
...
+        To do so, it reads first the content by using the file blob hash with the command `git cat-file -p [blob hash]`,
+        then it calculates the hash.
+        """
+        files_info: List[Tuple[FilePath, TagName, FilePath]] = []
+        current_dir = os.getcwd()
+        os.chdir(repo_dir_path)
+
+        for (file_path, tag_name, blob_hash) in files:
+            try:
+                file_content = subprocess.check_output(
```
Comment (polyedre) [inline lib/resources/git_resource.py:85]: Here you could add a comment explaining why subprocess is the right choice and why you don't use the Gitpython way of interacting with the command line.

File: lib/resources/git_resource.py [hunk 7]
```diff
@@ -27,90 +37,181 @@ def __init__(self, database: DbConnector):
...
+                )
+                file_hash = self.get_hash(file_content)
+                files_info.append((file_path, tag_name, file_hash))
+            except (ValueError, subprocess.CalledProcessError) as exception:
+                logger.error(exception)
+        os.chdir(current_dir)
+        return files_info
+
+    @staticmethod
+    def _get_diff_between_two_tags(tag_a: Tag, tag_b: Tag) -> List[Tuple[FilePath, TagName, BlobHash]]:
```
Comment (polyedre) [inline lib/resources/git_resource.py:97]: The name of the method does not seem to be accurate.

File: lib/resources/git_resource.py [hunk 8]
```diff
@@ -27,90 +37,181 @@ def __init__(self, database: DbConnector):
...
+
+        for tag in tags:
+            if tag.name == last_version:
+                return diff_versions
+            if tag.name == first_version:
+                can_track = True
+            if can_track is False:
+                continue
+            diff_versions.append(tag.name)
+        return diff_versions
```
Comment (polyedre) [inline lib/resources/git_resource.py:158]: You could use something like
```python
tag_names = list(map(lambda tag: tag.name, tags))
return tags[tag_names.index(first_version):tag_names.index(last_version)]
```

**Review by polyedre — CHANGES_REQUESTED** [review]
You can improve readability a bit more 😃

**Review by polyedre — APPROVED** [review]

File: lib/resources/git_resource.py [hunk 9]
```diff
@@ -2,19 +2,31 @@
...
 
-EXTENSION = ["txt", "css", "js"]
+FilePath = str
+BlobHash = str
+TagName = str
+FileHash = str
+GitFileMetadata = Tuple[FilePath, TagName, BlobHash]
+FileMetadata = Tuple[FilePath, TagName, FileHash]
+
+BLOB_MODE = 33188
```
Comment (polyedre) [inline lib/resources/git_resource.py:29]: What is the meaning if this value ? Can you add a comment here or use a more explicit name ? :)

**Review by polyedre — APPROVED** [review]

**Comment by fwininger** [comment]
@JulienTD en attente de rebase en local.

**Comment by fwininger** [comment]
@JulienTD toujours à rebase

**Pull request merged**